### PR TITLE
booking views

### PR DESCRIPTION
**Pull request by brendanator**
- **Refactor views into a module and a drag and drop test**
- **Support for adding extra args to extra views**
- **Basic month view**
- **Make the code a bit more readable**
- **Add some links**
- **add headdings and colours**
- **use built-in day list**
- **Basic day view**
- **Working drag and drop, does not save**
- **Update the booking on drag and drop**
- **Better formating for alpine.js**
- **the drag and the drop works**
- **Fix a potential none error**
- **Replace depracated function**
- **Fix clean slots command**
- **Fix booking slots not being cleaned up correctly**
- **margianly better styling**
- **size the grouped booking handler**
- **-m**
- **move the element into the booking group if there is one**
- **error toast**
- **Better error handling**
- **Change the error wording**
- **allow dropping on children**
- **Remove console.log**
- **I can compile typescript to js file!**
- **Minify the js**
- **I've already changed bundler, eslint is quicker and works on the cli properly**
- **Finished JS setup**
- **The js was getting a bit much for inline so move it and typescript it**
- **Make movebooking return a promis**
- **Convert it to async await**
- **Improve error handling**
- **Remove the unneeded htmx code**
- **Add a close button and max height**
- **Animate the toast in and out**
- **Tweak the animation**
- **Don't throw error on 400**
- **Simple year view, broken breadcrumbs**
- **tidy up the bread crumbs**
- **Fix breadcrumbs**
- **Add some stats and styling to the year view**
- **Fix the ajax nesting**
- **Remove dep on dateutil and fix shadowing**
- **Sort out the controles and update icons**
- **drag and drop month view**
- **fix url name**
- **Fix issue with overlaps**
- **show the time on the month view**
- **Moving bookings can be done with just a date and they will keep the same time**
- **Reuse more code and only the date when moveing between days**
- **improve year layout**
- **extend the day if there are bookings outside of the normal range**
- **add some padding**
- **Fix none timezone times**
- **custome make_awhere that doesn't error and accepts dates**
- **WIP: booking form with cost filled from service**
- **Select field that has data atrributes**
- **Use an actual float**
- **Move the widgets into a file**
- **If the value is function, call it**
- **change the cost based on the service selection, unless the cost has been changed**
- **Simple minifier for the alpine attributes**
- **Test the whitespace helper**
- **Allow setting of attributes in the form object**
- **Add age and name with owner**
- **Reduce the number of queries used in the data attr widget**
- **Always prefetch a pets customer**
- **recursive get attr function**
- **add the customer id to the pet dropdown**
- **Test rgettattr**
- **minimize whitespace doesn't in debug mode**
- **Linked inputs!**
- **Actualy linked form fields and select if there only one**
- **Swap logic for better readability**
- **It's only one customer**
- **Disable the table overflow**
- **Allow tables to overflow**
- **round the start and end time and perform some small manualy js minifaction**
- **Change alias name**
- **Move the alias into a file**
- **Add support for multipul data attributes**
- **Link the start and end together via the service length**
- **Set the data from the value**
- **Fix filling of default values**
- **Tidy up the single money widget**
- **Change the booking cost field a moeny field**
- **Move some code to a function**
- **Fix details view and urls**
- **don't let the same pet on a booking slot more than once**
- **Change the constriantins so I canceled bookings can have no slot**
- **Fix tests**
- **Move constraints to the db layer**
- **Test new constraints**
- **Actualy apply the constraints and remove tests nolonger needed**
- **Move the booking status outside of booking so I can access them in the meta class**
- **Remove dummy data requirement from makefile**
- **Add a db constraint to the state field**
- **Active booking manger so the calenders only show active**
- **Add booking fabs to the booking pages**
- **Fix moveing of booking state**
- **Allow cost per additional to be none**
- **show the customer on the booking details**
- **Fill the customer if the pet is set and reorder the inputs**
- **Add overload and booking slot validation**
- **Check the booking slot has an id**
- **TODO: Handle dropping groups ontop of groups**
- **Fix timezone issue, how can you not just set a datetime object into a datetime input?**
- **if the container is the same as element, move the children and remove the container**
- **Left and right arrows navigate the calendar**
- **fix issue with drop**
- **Add grip texture**
- **Issue an even on booking move and reload the time slots**
- **Fix higlighing of today**
- **Set draggable corectly and auto pointer**
- **WIP: change the bookings from pet based to customer based**
- **WIP: more moving bookings to customers**
- **fix the booking titles**
- **Convert the data attr fields to mixins**
- **WIP: convert the booking form**
- **WIP: fix more changes with bookings**
- **WIP: ensure pets and customers match on bookings**
- **Update and fix tests**
- **Fix the last of the tests**
- **Add a can move to a booking slot and fix the group drag**
- **Better handling of no selectable pets**
- **Simplfy the widgets the auto select the pets after selecting the customer**
- **Fix creating a booking**
- **Add a detail link to bookings in the day view**
- **make the breadcrumb ordonal**
- **Move the icons to static files and add a template tag to include static files**
- **Swap the css gradient for an svg**
- **Ensure the fab is above everything**
- **Missed an icon**
- **Create and update views catch value and validations errors in the save process**
- **Don't allow minus values**
- **Ensure the pets belong to the customer**
- **Add a field for additonaly cost**
- **Add the additioal to details**
- **Generate the name field**
- **Rename wire up the aditonal field**
- **Update how the charges are created**
- **Add some links to the details page**
- **WIP: form for invoicing charges**
- **Basic widget to display checkboxes and there info in a table**
- **Ensure there is a charge created when there are no pets assigned**
- **Mark the the test as untestable**
- **python 3.11 support**
- **More python 3.11 support**
- **Add coverage**
- **Add pytz**
- **Improve test coverage**
- **select/unselect all items**
- **Better form layout and stop sort erroring**
- **Fix sorting when there are no filters**
- **Disabled select all when there no options**
- **Remove dummy table**
- **Add uninvoiced count and marker**
- **generate the circles using cog**
- **add a gap and tidy the css output**
- **Better formatting**
- **Use correct colours**
- **WIP: charge form submission**
- **Only show unpaid if that are not overdue**
- **Remove comment**
- **show the dots as a background image to avoid distortion**
- **Remove test classes**
- **swap back to content**
- **put the size in the SVG**
- **working charge form**
- **Apply linting**
- **Version upgrades**
- **None ajax version of the charge form**
- **Move the icons back**
- **Invoice actions and icons**
- **move the type ignore**
- **WIP: add a step with a form for transitions**
- **conformation form for sending an invoice**
- **Fix dialog closing**
- **Fix some dialog styling**
- **Update wording**
- **Reduce the height of the text boxes**
- **Support for post with variables for the transition view**
- **use the correct urls for the send form**

### CODE REVIEW

**Comment by brendanator** [comment]
This is a benchmark review for experiment `review_of_100_reviews_20240409`.
Run ID: `review_of_100_reviews_20240409/benchmark_2024-04-09T10-24-34_v1-16-0-95-g65fefaeef-dirty`.

This pull request was cloned from `https://github.com/bengosney/cerberus/pull/788`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

File: cerberus/models/booking.py [hunk 1]
```diff
@@ -57,38 +69,60 @@ def round_date_time(dt: datetime) -> datetime:
...
-        return self.end > self.start
-
     def get_overlapping(self) -> QuerySet[Self]:
         start = Q(start__lt=self.start, end__gt=self.start)
         end = Q(start__lt=self.end, end__gt=self.end)
         equal = Q(start=self.start, end=self.end)
+        contains = Q(start__gt=self.start, end__lt=self.end)
 
-        return self.__class__.objects.filter(start | end | equal).exclude(pk=self.pk)  # type: ignore
+        return self.__class__.objects.filter(start | end | equal | contains).exclude(pk=self.pk)
```
Comment (brendanator) [inline cerberus/models/booking.py:78]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (code_refinement):** Consider simplifying the overlap check logic.

The current implementation of the `overlaps` method might be simplified by directly using Django's `__overlap` lookup in the query. This could potentially improve readability and performance.

```suggestion
def get_overlapping(self) -> QuerySet[Self]:
    return self.__class__.objects.filter(
        Q(start__lt=self.end, end__gt=self.start)
    ).exclude(pk=self.pk)
```

File: cerberus/models/booking.py [hunk 2]
```diff
@@ -57,38 +69,60 @@ def round_date_time(dt: datetime) -> datetime:
...
+        if (max_customer := getattr(self.service, "max_customer")) and self.customer_count > max_customer:
+            raise ValidationError(f"{self.__class__.__name__} has max customers for service, {max_customer}")
+
+        super().clean()
+
+    @property
+    def can_move(self) -> bool:
+        return all(b.can_move for b in self.bookings.all())
+
+    def move_slot(self, start: datetime | date, end: datetime | None = None) -> bool:
```
Comment (brendanator) [inline cerberus/models/booking.py:104]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (edge_case_not_handled):** Validate the return value of `move_slot` method.

The method always returns `True`, which might not accurately reflect the outcome of the operation. Consider implementing a mechanism to return `False` when the move operation is not successful.

```suggestion
        if not self.can_move:
            return False

        # Your existing logic to move the slot

        # After moving logic
        if self.get_overlapping().exists():  # Assuming get_overlapping() checks for overlaps after moving
            return False  # Move was not successful due to overlaps

        return True  # Move was successful
```

File: cerberus/models/booking.py [hunk 3]
```diff
@@ -187,47 +286,99 @@ def save(self, *args, **kwargs) -> None:
...
+                cost = self.cost_per_additional
+                if self.cost_per_additional is None:
+                    break
+        else:
+            name = f"{self.service}"[:255]
+            charge = BookingCharge(name=name, line=cost, booking=self, customer=self.customer)
+            charge.save()
+            charges.append(charge)
+
+        return charges
```
Comment (brendanator) [inline cerberus/models/booking.py:363]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (performance):** Consider bulk creation of `Charge` objects.

Creating `Charge` objects in a loop might not be the most efficient approach. Explore using `bulk_create` for better performance.

File: cerberus/models/service.py [hunk 4]
```diff
@@ -33,3 +33,9 @@ def __str__(self) -> str:
 
     def get_absolute_url(self) -> str:
         return reverse("service_detail", kwargs={"pk": self.pk})
+
+    def length_seconds(self) -> int:
+        return int(self.length.total_seconds())
+
+    def length_minutes(self) -> int:
+        return self.length_seconds() // 60
```
Comment (brendanator) [inline cerberus/models/service.py:41]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (code_refinement):** Redundant method for calculating length in minutes.

Given that `length` is already a `DurationField`, consider using Django's built-in methods to get the duration in minutes directly, rather than converting seconds to minutes manually.

File: cerberus/widgets.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        selected: set[str] | bool,
+        index: int,
+        subindex: int | None = None,
+        attrs: dict[str, Any] | None = None,
+    ) -> dict[str, Any]:
+        option = super().create_option(name, value, label, selected, index, subindex, attrs)
+        if self.attr_callback:
+            option["attrs"] = self.attr_callback(name, value, label, option["attrs"])
+
+        return option
```
Comment (brendanator) [inline cerberus/widgets.py:56]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (performance):** Consider optimizing CheckboxTable's create_option method for large datasets.

The current implementation iterates over all model fields for each option, which might be inefficient for large datasets or complex models. Consider caching or pre-fetching data if possible.

File: cerberus/views/invoice.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        if request.POST and form.is_valid():
+            invoice.send(form.cleaned_data["to"], form.cleaned_data["send_email"])
+            return HttpResponseRedirect(reverse_lazy("invoice_detail", kwargs={"pk": invoice.pk}))
+
+        return render(request, "cerberus/invoice_send_form.html", {"invoice": invoice, "form": form})
+
+
+class InvoiceActionsView(TransitionView):
+    model = Invoice
+    field = "state"
```
Comment (brendanator) [inline cerberus/views/invoice.py:128]: <a href="https://sourcery.ai">Sourcery logo</a>  **question (code_refinement):** Review the necessity of InvoiceActionsView inheriting from TransitionView.

Ensure that the inheritance from TransitionView is necessary and that the functionality aligns with the intended use case for InvoiceActionsView.

File: cerberus/tests/test_bookings.py [hunk 7]
```diff
@@ -155,21 +134,22 @@ def test_pet_count(walk_service, make_pet):
...
+        pets=[make_pet()],
     ).booking_slot
 
     assert booking_slot_1.pk == booking_slot_2.pk
     assert booking_slot_1.pet_count == 2
 
 
 @pytest.mark.django_db
-def test_customer_count(walk_service, make_pet, customer):
+def test_customer_count(walk_service):
```
Comment (brendanator) [inline cerberus/tests/test_bookings.py:145]: <a href="https://sourcery.ai">Sourcery logo</a>  **suggestion (testing):** Consider adding more context or a docstring to `test_customer_count`.

It's not immediately clear what specific behavior or edge case this test is verifying. Adding a brief explanation could improve readability.

File: cerberus/tests/test_utils.py [hunk 8]
```diff
@@ -13,3 +22,127 @@ class TestLength(models.TextChoices):
         __empty__ = "empty", "Empty"
 
     assert choice_length(TestLength) == 5
+
+
+def test_choice_length():
```
Comment (brendanator) [inline cerberus/tests/test_utils.py:27]: <a href="https://sourcery.ai">Sourcery logo</a>  **issue (testing):** Duplicate test function `test_choice_length`.

This function appears to be a duplicate of an earlier test function for `choice_length`. Consider removing or renaming it if it serves a different purpose.

File: cerberus/forms.py [hunk 9]
```diff
@@ -77,18 +54,128 @@ class Meta:
 
 
 class BookingForm(forms.ModelForm):
+    attributes = {
```
Comment (brendanator) [inline cerberus/forms.py:57]: <a href="https://sourcery.ai">Sourcery logo</a>  **issue (complexity):** Consider separating frontend logic from Django forms to maintain a clear separation of concerns.

The introduction of frontend logic directly into Django forms, particularly through attributes like `x-data`, `x-model.number.fill`, `@change`, etc., significantly increases the complexity of the codebase. This approach mixes backend form logic with frontend behavior, leading to a blurring of the separation of concerns. It makes the code harder to maintain, as developers now need to navigate both Python and JavaScript within the same files. Additionally, the use of custom widgets and attributes for frontend interactions further complicates the form definitions.

A more maintainable approach would be to keep the frontend logic separate from the Django forms. Instead of embedding JavaScript directly into the form widgets, consider using data attributes minimally to mark elements for frontend interaction and handling the behavior entirely in separate JavaScript files. This maintains a clear separation of concerns, simplifies the form definitions, and reduces the cognitive load on developers by keeping backend and frontend logic distinct.

File: cerberus/models/pet.py [hunk 10]
```diff
@@ -21,6 +21,11 @@
     from . import Booking
 
 
+class PetManager(models.Manager):
```
Comment (brendanator) [inline cerberus/models/pet.py:24]: <a href="https://sourcery.ai">Sourcery logo</a>  **issue (complexity):** Consider simplifying the Pet model by reducing external dependencies and custom manager complexity.

While the enhancements introduced in this PR, such as the custom `PetManager` and the use of `humanize.naturaldelta` for calculating the pet's age, aim to improve the functionality and user experience, they also add a notable layer of complexity to the `Pet` model. The custom manager, while powerful, might be an over-engineering for the current requirements, especially if the performance gains are minimal with the existing dataset. Similarly, the dependency on `humanize.naturaldelta` for a more readable age format introduces an external library dependency that needs to be maintained and understood.

Moreover, the addition of model properties like `name_with_owner` and `age` increases the model's responsibilities, potentially making it harder to maintain as more logic is added. It's generally beneficial to keep models lean and focused primarily on representing the database structure, with business logic residing in services or utility functions.

To simplify, consider calculating the pet's age directly within the model without an external library. This approach reduces complexity and external dependencies while still providing a clear, human-readable age format. Here's a suggested implementation:

```python
from datetime import datetime

class Pet(models.Model):
    # Existing fields...

    @property
    def age(self) -> str:
        if self.dob:
            delta = datetime.now().date() - self.dob
            years = delta.days // 365
            months = (delta.days % 365) // 30
            if years > 0:
                return f"{years} years, {months} months"
            else:
                return f"{months} months"
        return "Unknown"
```

This method directly calculates the age in a straightforward manner, avoiding additional complexities and dependencies. It's a more maintainable approach that keeps the model's codebase lean and focused.

File: cerberus/urls/urls.py [hunk 11]
```diff
@@ -6,8 +6,46 @@
 
 urlpatterns = (
     [
-        path("", views.dashboard, name="dashboard"),
-        path("bookings/<int:pk>/action/<str:action>", views.BookingStateActions.as_view(), name="booking_action"),
+        path(
```
Comment (brendanator) [inline cerberus/urls/urls.py:9]: <a href="https://sourcery.ai">Sourcery logo</a>  **issue (complexity):** Consider maintaining the original single-line format for simpler paths to enhance readability and manage complexity.

The refactoring introduces additional complexity and visual clutter by significantly increasing the line count and breaking down each `path` function call into multiple lines. This not only makes the `urlpatterns` list longer but also introduces a level of repetition, especially with the booking calendar views (year, month, day), that could be abstracted to reduce both repetition and complexity.

While the intention to improve readability is appreciated, the original single-line format for each path was more consistent and, for many, easier to quickly scan through. The addition of new features (calendar views and invoice actions) inherently increases complexity, but the way these new paths are interspersed with existing ones without clear separation or grouping could make the codebase harder to maintain.

Consider maintaining the original formatting for simpler paths and only using multi-line formatting for more complex paths if it genuinely improves readability. Additionally, grouping related paths or abstracting similar patterns could help manage complexity while keeping the codebase clean and maintainable.

File: cerberus/forms.py [hunk 12]
```diff
@@ -1,33 +1,10 @@
...
-                }
-            ),  # type: ignore
-            currency_widget=forms.HiddenInput(),
-            *args,
-            **kwargs,
-        )
-
-    def id_for_label(self, id_):
-        return f"{id_}_0"
+from .utils import minimize_whitespace
```
Comment (brendanator) [inline cerberus/forms.py:6]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Consider removing unused import.

The import 'minimize_whitespace' from '.utils' does not seem to be used in 'forms.py'. If it's not needed, consider removing it to keep the code clean.

File: cerberus/api.py [hunk 13]
```diff
@@ -11,7 +11,6 @@
 from django.urls import path
 
 # Third Party
-from dateutil import parser
```
Comment (brendanator) [inline cerberus/api.py:14]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Good job on removing unused imports.

Removing the unused 'parser' import from 'dateutil' helps in keeping the codebase clean and reduces unnecessary dependencies.

File: cerberus/api.py [hunk 14]
```diff
@@ -125,31 +124,35 @@ def complete(self, request, pk=None):
         return self.change_state("complete", request)
 
     @action(detail=True, methods=["PUT"])
```
Comment (brendanator) [inline cerberus/api.py:127]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Ensure API endpoint renaming is reflected in the frontend.

The renaming of the 'move_booking' method to 'move' changes the API endpoint. Make sure this change is updated in any frontend code that calls this endpoint to avoid breaking the application.

File: cerberus/api.py [hunk 15]
```diff
@@ -125,31 +124,35 @@ def complete(self, request, pk=None):
...
-    def move_booking(self, request, pk=None):
-        booking = self.get_object()
```
Comment (brendanator) [inline cerberus/api.py:129]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Verify serializer replacement maintains functionality.

Replacing 'BookingMoveSerializer' with 'ToDateSerializer' may affect data validation and parsing. Ensure that the new serializer adequately supports the required fields and validation rules for booking movement.

File: cerberus/api.py [hunk 16]
```diff
@@ -125,31 +124,35 @@ def complete(self, request, pk=None):
...
+                if booking.move_booking(incoming.validated_data["to"]):
                     status = 200
 
         serializer = self.get_serializer(booking)
 
         return Response({"item": serializer.data, "status": status}, status=status)
 
     @action(detail=True, methods=["PUT"])
-    def move_booking_slot(self, request, pk=None):
-        booking = self.get_object()
```
Comment (brendanator) [inline cerberus/api.py:145]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Ensure consistency in API endpoint naming.

Renaming 'move_booking_slot' to 'move_slot' improves consistency with the 'move' method for bookings. Verify that all related documentation and frontend calls are updated accordingly.

File: cerberus/utils.py [hunk 17]
```diff
@@ -1,6 +1,38 @@
...
+    if settings.DEBUG:
+        return value
+    return re.sub(r"(^\s+|[\n\r]+)", "", value, flags=re.MULTILINE).strip()
+
+
+def rgetattr(obj, attr, *args):
+    def _getattr(obj, attr):
+        return getattr(obj, attr, *args)
+
+    return functools.reduce(_getattr, [obj] + attr.split("."))
```
Comment (brendanator) [inline cerberus/utils.py:38]: <a href="https://sourcery.ai">Sourcery logo</a>  **🚨 issue (security):** Review new utility functions for potential overlaps.

The addition of 'make_aware', 'minimize_whitespace', and 'rgetattr' utility functions is noted. Ensure there's no overlap with existing Django utilities or third-party libraries to avoid redundancy.

**Review by brendanator — COMMENTED** [review]
<a href="https://sourcery.ai">Sourcery logo</a>  Sourcery Review:

Hey @brendanator - I've reviewed your changes and found some issues that need to be addressed.

**Blocking issues**:
- Consider removing unused import. ([link](https://github.com/sourcery-ai-experiments/cerberus/pull/1/files#diff-4ab36e472b47ef5939ee308e54a945667a3f8cd5947e9b4a8321d39d95311444R6))
- Good job on removing unused imports. ([link](https://github.com/sourcery-ai-experiments/cerberus/pull/1/files#diff-df9cff77e0677dee368da0c3775f7cac0ec5f174b4d245ad14c5fe04dadd7bf5L14))
- Ensure API endpoint renaming is reflected in the frontend. ([link](https://github.com/sourcery-ai-experiments/cerberus/pull/1/files#diff-df9cff77e0677dee368da0c3775f7cac0ec5f174b4d245ad14c5fe04dadd7bf5L128))
- Verify serializer replacement maintains functionality. ([link](https://github.com/sourcery-ai-experiments/cerberus/pull/1/files#diff-df9cff77e0677dee368da0c3775f7cac0ec5f174b4d245ad14c5fe04dadd7bf5L130))
- Ensure consistency in API endpoint naming. ([link](https://github.com/sourcery-ai-experiments/cerberus/pull/1/files#diff-df9cff77e0677dee368da0c3775f7cac0ec5f174b4d245ad14c5fe04dadd7bf5L144))
- Review new utility functions for potential overlaps. ([link](https://github.com/sourcery-ai-experiments/cerberus/pull/1/files#diff-193e01732e14dfc63e48b2848dcb8dc1246bd4b052b47b55c46953f7a09c5940R17))

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 6 issues found
- 🔴 **Security**: 6 blocking issues
- 🟡 **Testing**: 2 issues found
- 🟡 **Complexity**: 3 issues found
- 🟢 **Docstrings**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/5c8bea2b-6c6c-4221-bd9b-0c11556933a8/projects/p/2dbf7697-3bb3-4f63-b184-3a8ebf7ac23c/r/428ffbd5-7b47-4bad-8de0-81731e485bc9?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

File: cerberus/models/booking.py (hunk 1 above)
Comment (SourceryAI) [reply cerberus/models/booking.py:78]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/models/booking.py:78]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/models/booking.py:78]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/models/booking.py:78]: Is the comment area correct?

File: cerberus/models/booking.py (hunk 2 above)
Comment (SourceryAI) [reply cerberus/models/booking.py:104]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/models/booking.py:104]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/models/booking.py:104]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/models/booking.py:104]: Is the comment area correct?

File: cerberus/models/booking.py (hunk 3 above)
Comment (SourceryAI) [reply cerberus/models/booking.py:363]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/models/booking.py:363]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/models/booking.py:363]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/models/booking.py:363]: Is the comment area correct?

File: cerberus/models/service.py (hunk 4 above)
Comment (SourceryAI) [reply cerberus/models/service.py:41]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/models/service.py:41]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/models/service.py:41]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/models/service.py:41]: Is the comment area correct?

File: cerberus/widgets.py (hunk 5 above)
Comment (SourceryAI) [reply cerberus/widgets.py:56]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/widgets.py:56]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/widgets.py:56]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/widgets.py:56]: Is the comment area correct?

File: cerberus/views/invoice.py (hunk 6 above)
Comment (SourceryAI) [reply cerberus/views/invoice.py:128]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/views/invoice.py:128]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/views/invoice.py:128]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/views/invoice.py:128]: Is the comment area correct?

File: cerberus/tests/test_bookings.py (hunk 7 above)
Comment (SourceryAI) [reply cerberus/tests/test_bookings.py:145]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/tests/test_bookings.py:145]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/tests/test_bookings.py:145]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/tests/test_bookings.py:145]: Is the comment area correct?

File: cerberus/tests/test_utils.py (hunk 8 above)
Comment (SourceryAI) [reply cerberus/tests/test_utils.py:27]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/tests/test_utils.py:27]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/tests/test_utils.py:27]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/tests/test_utils.py:27]: Is the comment area correct?

File: cerberus/forms.py (hunk 9 above)
Comment (SourceryAI) [reply cerberus/forms.py:57]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/forms.py:57]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/forms.py:57]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/forms.py:57]: Is the comment area correct?

File: cerberus/models/pet.py (hunk 10 above)
Comment (SourceryAI) [reply cerberus/models/pet.py:24]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/models/pet.py:24]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/models/pet.py:24]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/models/pet.py:24]: Is the comment area correct?

File: cerberus/urls/urls.py (hunk 11 above)
Comment (SourceryAI) [reply cerberus/urls/urls.py:9]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/urls/urls.py:9]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/urls/urls.py:9]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/urls/urls.py:9]: Is the comment area correct?

File: cerberus/forms.py (hunk 12 above)
Comment (SourceryAI) [reply cerberus/forms.py:6]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/forms.py:6]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/forms.py:6]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/forms.py:6]: Is the comment area correct?

File: cerberus/api.py (hunk 13 above)
Comment (SourceryAI) [reply cerberus/api.py:14]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/api.py:14]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/api.py:14]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/api.py:14]: Is the comment area correct?

File: cerberus/api.py (hunk 14 above)
Comment (SourceryAI) [reply cerberus/api.py:127]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/api.py:127]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/api.py:127]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/api.py:127]: Is the comment area correct?

File: cerberus/api.py (hunk 15 above)
Comment (SourceryAI) [reply cerberus/api.py:129]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/api.py:129]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/api.py:129]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/api.py:129]: Is the comment area correct?

File: cerberus/api.py (hunk 16 above)
Comment (SourceryAI) [reply cerberus/api.py:145]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/api.py:145]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/api.py:145]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/api.py:145]: Is the comment area correct?

File: cerberus/utils.py (hunk 17 above)
Comment (SourceryAI) [reply cerberus/utils.py:38]: Is this comment correct?

Comment (SourceryAI) [reply cerberus/utils.py:38]: Is this comment helpful?

Comment (SourceryAI) [reply cerberus/utils.py:38]: Is the comment type correct?

Comment (SourceryAI) [reply cerberus/utils.py:38]: Is the comment area correct?
